### DISCUSSION TITLE
Why find-debuginfo script only operates on executable files?

### DISCUSSION THREAD

**Discussion by sshedi** [rpm-software-management/rpm#3195] [Q&A] [answered]
Previously find-debuginfo was a part of RPM, so I hope it's okay to ask this question here.

I want to know the history behind why find-debuginfo.sh operates only on exec files.

https://sourceware.org/git/?p=debugedit.git;a=blob;f=scripts/find-debuginfo.in;h=ae0818fb6bfc8ef96bc3f24c3156455d50c91506;hb=HEAD#l437

This makes things a bit difficult while generating debuginfo package of kernels.
`make modules_install` will install all `.ko` (kernel objects) under `<buildroot>/lib/modules` but find-debuginfo won't operate on it due to 644 permissions. Is there any reason to operate only on exec files? Why can't we operate on all ELF files?

**Comment by sshedi** [comment]
Hi @pmatilai, any inputs on the above issue?
I faced a similar issue with another package recently. My package has only few shared object files and I installed those `.so` files with 644 permission in RPM_BUILD_ROOT and find-debuginfo failed with empty debugfiles.list error. But newer version of rpm changes .so file permissions to 644 by default and I installed them as 644 by default.

RPM post install script might do the permission changes post gathering debuginfo data but installing .so files with 644 permission is a valid case. Can you please provide some info. I need to know if this should be handled at individual spec level or can be fixed in find-debuginfo tool itself. Thanks.

**Comment by dmnks** [comment]
(Just FYI, @pmatilai is on vacation until August.)

This is a convention (at least on Linux) for shared libraries to be executable. It's also [mandated](https://docs.fedoraproject.org/en-US/packaging-guidelines/Debuginfo/#_checking_your_debuginfo_package_for_usefulness) by Fedora Packaging Guidelines. I'm not familiar with the actual reasoning behind it but a quick google search gives some [clues](https://unix.stackexchange.com/a/61650).

As to why find-debuginfo.sh only looks for those that *are* executable: Looking at the git log, this check apparently has been in place since the very first version of the script added in 2002 (commit 90b34cd4e51de9b69a1a1e4bf58fcee31c2bf192). That said, the original brp-strip-shared script (no longer included, see commit fd519223ae91f254a913c26f09c1289bf687e277), on the other hand, does *not* have this filter (and instead mentions in a cryptic comment that "our libraries aren't +x"). So that's quite confusing.

In any case, you may want to ask about this at https://sourceware.org/debugedit/ (which is where find-debuginfo.sh is nowadays shipped). I personally don't think anybody would change this now (without a very good reason) after those 20+ years. Even if the check itself turns out to be non-sensical, it should be considered a "feature" at this point because changing it would inevitably break somebody's packages, possibly in non-obvious or undetected ways.

**Comment by sshedi** [comment]
Thanks for the background and insights @dmnks.

@pmatilai, please let us know if you have any further points on this. Or else I will proceed and close the discussion.
Thanks.

**Comment by pmatilai** [comment]
Enabling brp-elfperms (which disables the executable bits in recent rpm versions) seems to be an error to begin with, at least on modern Linux. According to @fweimer all shared objects *must* be executable: https://lists.rpm.org/pipermail/rpm-list/2023-March/002093.html

It also gets it wrong sometimes: https://github.com/rpm-software-management/rpm/issues/2496
Since the whole thing seems misguided (at least on Linux), we should probably just disable it by default...

As for why debuginfo mechanism is tied to it, probably "historical" (the easiest ballpark estimate of what to process) to begin with, and also used as a heuristic for dependency generation back then. But that's just a guess, this all predates my involvement by several years. It's also pretty much the only chance for packages to control debuginfo behavior per-file.

**Comment by sshedi** [comment]
Thanks for the response @pmatilai, if I have understood it right `brp-elfperms` will stop removing executable permission bit from shared objects in a near future version of rpm? Meaning, reverting back to older mechanism of shipping .so files with executable bit set.

**Comment by pmatilai** [comment, marked as answer]
The raison d'etre of brp-elfperms is to strip permissions from files that don't need it, so it's more like the whole script should just be disabled. If you build + ship your own rpm on Linux, then I'd recommend you do just that. For the time being I mean.

**Comment by sshedi** [comment]
Noted, thanks for all the inputs and suggestions. Have a nice time ahead.
